### PR TITLE
Maximum number of members

### LINKED ISSUE #288: Group Assignment - Maximum number of members
This is a feature request, to add an optional field in the settings for the maximum number of members per team.

This is a setting would either: only be set on initial creation and be immutable afterwards, or could only be edited to a higher value.
Otherwise, if the value could be edited to a lower number of members per team, there would be the problems of removing members from a team, which members would be removed, and where those members would go.

### PR DESCRIPTION
**Pull request by dinever**
Hi guys,

This PR fixes #288.

Besides adding the maximum members feature, I also modified the group assignment invitation page. Now it has a better look and makes it easier for students to choose the correct team:

screen shot 2016-03-22 at 1 09 06 am

Any suggestions & feedback will be greatly appreciated. :smile:

### CODE REVIEW

File: spec/controllers/group_assignment_invitations_controller_spec.rb [hunk 1]
```diff
@@ -22,7 +22,7 @@
                              title: 'HTML5',
                              grouping: grouping,
                              organization: organization,
-                             public_repo: false)
+                             public_repo: true)
```
Comment (dinever) [inline spec/controllers/group_assignment_invitations_controller_spec.rb]: I wonder what was the reason for us to use private repo in the previous test cases.

I changed it to `true` as I was not able to create a private repo in a GitHub organization and can not record a VCR of creating a private repo.

Maybe I can figure it out later and fix this.

Comment (tarebyte) [inline spec/controllers/group_assignment_invitations_controller_spec.rb]: It was because I was using my personal token and I didn't want the creation to show up on my timeline.

No biggie

File: app/controllers/group_assignment_invitations_controller.rb [hunk 2]
```diff
@@ -7,6 +7,7 @@ class GroupAssignmentInvitationsController < ApplicationController
   before_action :authorize_group_access, only: [:accept_invitation]
 
   def show
+    @group_assignment_repos = invitation.group_assignment.group_assignment_repos
```
Comment (tarebyte) [inline app/controllers/group_assignment_invitations_controller.rb]: We already have an `#group_assignment` method, could you update it to use that?

**Comment by tarebyte** [comment]
@dinever I like this style a lot, what do we think about making this the way we also list groups for the assignment as well.

/cc @jasonlong for thoughts

File: db/schema.rb [hunk 3]
```diff
@@ -92,6 +92,7 @@
     t.integer  "creator_id"
     t.datetime "deleted_at"
     t.string   "slug",                                null: false
+    t.integer  "max_members"
```
Comment (tarebyte) [inline db/schema.rb]: For production applications we're going to want to extract the database migration as it's own PR and we'll deploy and run the migration before we merge this PR.

@dinever if you need any assistance with that please let me know.

**Comment by tarebyte** [comment]
@dinever this looks really good to me, if you can follow up with a migration PR I talked about here https://github.com/education/classroom/pull/511#discussion_r57091591 it'd

**Comment by dinever** [comment]
Thanks! @tarebyte 

Just had the code modified according to the comments. The migration PR is at #512 now.

**Comment by jasonlong** [comment]
:thumbsup: LGTM!

File: app/controllers/group_assignment_invitations_controller.rb [hunk 4]
```diff
@@ -52,11 +52,19 @@ def authorize_group_access
...
 
     return unless group_id.present?
+    group = Group.find(group_id)
+    check_group_members_not_exceed(group)
     return if group_assignment.grouping.groups.find_by(id: group_id)
 
     fail NotAuthorized, 'You are not permitted to select this team'
   end
 
+  def check_group_members_not_exceed(group)
```
Comment (tarebyte) [inline app/controllers/group_assignment_invitations_controller.rb]: what do you think about `#validate_group_member_limit_not_exceeded!`

Comment (johndbritton) [inline app/controllers/group_assignment_invitations_controller.rb]: let's keep the naming consistent: `#validate_max_members_not_exceeded`

@tarebyte does it make sense to have `!` here?

File: app/views/group_assignment_invitations/_group_assignment_team.erb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+          <% end %>
+      <% end %>
+      <% if group_member_count > 10 %>
+        <%= link_to group_assignment_repo.github_team_url, class: "#{'disabled' if group_assignment_repo.disabled?}" do %>
+          <%= image_tag 'student-ellipse@2x.png', class: 'avatar avatar-small', height: 30, width: 30, alt: 'student ellipse' %>
+        <% end %>
+      <% end %>
+    </div>
+  </div>
+</div>
```
Comment (johndbritton) [inline app/views/group_assignment_invitations/_group_assignment_team.erb]: Missing newline at eof.

Comment (johndbritton) [inline app/views/group_assignment_invitations/_group_assignment_team.erb]: Can we make the build fail on missing newline @tarebyte?

cc/ https://github.com/education/classroom/issues/514

File: app/controllers/group_assignment_invitations_controller.rb (hunk 4 above)
Comment (tarebyte) [inline app/controllers/group_assignment_invitations_controller.rb]: @johndbritton it does because we raise an error

File: app/views/group_assignment_invitations/show.html.erb [hunk 6]
```diff
@@ -36,7 +36,9 @@
       <dt><label><%= 'OR' if @groups.present? %> Create a new team</label></dt>
       <dd>
         <%= text_field_tag 'group[title]', nil, placeholder: 'Create a new team', class: 'textfield' %>
-        <%= submit_tag 'Create team', class: 'btn btn-green js-navigation js-form-submit', disabled: true %>
+        <%= button_tag(type: 'submit', class: 'btn btn-green js-navigation js-form-submit') do %>
+          <i class="octicon octicon-plus"></i> Create Team
```
Comment (tarebyte) [inline app/views/group_assignment_invitations/show.html.erb]: So we have an octicon helper just for occasions like these.

```erb
<%= octicon('plus') %> Create team
```

If you could also lower case `Team` that would be ✨

File: app/views/group_assignment_invitations/show.html.erb [hunk 7]
```diff
@@ -19,21 +19,27 @@
...
+  <hr>
 
+  <%= form_tag({ controller: :group_assignment_invitations, action: :accept_invitation}, method: :patch) do %>
     <dl class="form">
       <dt><label><%= 'OR' if @groups.present? %> Create a new team</label></dt>
-      <dd><%= text_field_tag 'group[title]', nil, placeholder: 'Create a new team', class: 'textfield' %></dd>
+      <dd>
+        <%= text_field_tag 'group[title]', nil, placeholder: 'Create a new team', class: 'textfield' %>
+        <%= button_tag(type: 'submit', class: 'btn btn-green js-navigation js-form-submit') do %>
+          <i class="octicon octicon-plus"></i> Create Team
```
Comment (tarebyte) [inline app/views/group_assignment_invitations/show.html.erb]: So we have an octicon helper just for occasions like these.

```erb
<%= octicon('plus') %> Create team
```

If you could also lower case `Team` that would be ✨

**Comment by tarebyte** [comment]
This looks good to ship, thanks @dinever!

**Comment by dinever** [comment]
Thanks @tarebyte,

I will consider adding this style to the group assignment students area as well in the future.

File: app/views/group_assignment_invitations/_group_assignment_team.erb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        <%= hidden_field_tag 'group[id]', group_assignment_repo.id %>
+        <% if group_member_count >= group_assignment.max_members %>
+            <%= submit_tag 'Full', class: 'btn', disabled: true %>
+        <% else %>
+            <%= submit_tag 'Join', class: 'btn' %>
+        <% end %>
+    <% end %>
+
+    <span class="team-name" itemprop="name"><%= group_assignment_repo.team_name %></span>
+    <p class="team-description" itemprop="description"><%=group_member_count%> <% if group_member_count != 1%>members<% else %>member<% end %></p>
```
Comment (johndbritton) [inline app/views/group_assignment_invitations/_group_assignment_team.erb]: `<% if group_member_count != 1%>members<% else %>member<% end %>`

Check out the [pluralize helper](http://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-pluralize).

File: app/views/group_assignment_invitations/_group_assignment_team.erb [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    <div class="team-members">
+      <% group_assignment_repo.group.repo_accesses.limit(7).each do |repo_access| %>
```
Comment (johndbritton) [inline app/views/group_assignment_invitations/_group_assignment_team.erb]: This is limited to `7` here and you're using `10` down below, why the difference?

Comment (dinever) [inline app/views/group_assignment_invitations/_group_assignment_team.erb]: @johndbritton Sorry for this mistake. Just got them fixed & improved.

**Comment by johndbritton** [comment]
This is looking pretty good. Just did a pass and I think the views could be improved by extracting the logic that's happening there into the controller and sending the data through instance variables.

@tarebyte I think this is another case where view models would be helpful.

File: spec/controllers/group_assignment_invitations_controller_spec.rb [hunk 10]
```diff
@@ -58,6 +58,24 @@
         expect(group_assignment.group_assignment_repos.count).to eql(0)
         expect(user.repo_accesses.count).to eql(0)
       end
+
+      it 'does not allow user to join if group has reached maximum number of members' do
```
Comment (johndbritton) [inline spec/controllers/group_assignment_invitations_controller_spec.rb]: You should use a context here:

```ruby
context 'group has reached maximum number of members' do
  it 'does not allow a user to join' do
   ...
  end
end
```

File: spec/controllers/group_assignment_invitations_controller_spec.rb [hunk 11]
```diff
@@ -58,6 +58,24 @@
...
+        group = Group.create(title: 'The Group', grouping: grouping)
+        group_assignment.update(max_members: 1)
+        student = GitHubFactory.create_classroom_student
+        group.repo_accesses << RepoAccess.create(user: student, organization: organization)
+
+        expect_any_instance_of(ApplicationController).to receive(:flash_and_redirect_back_with_message)
+        patch :accept_invitation, id: invitation.key, group: { id: group.id }
+      end
+
+      it 'allows user to join a group if maximum number of members is not set' do
```
Comment (johndbritton) [inline spec/controllers/group_assignment_invitations_controller_spec.rb]: Same as above

File: spec/controllers/group_assignment_invitations_controller_spec.rb (hunk 10 above)
Comment (johndbritton) [inline spec/controllers/group_assignment_invitations_controller_spec.rb]: You should include the test setup in the `context` and the actual test run in the `it`

Comment (johndbritton) [inline spec/controllers/group_assignment_invitations_controller_spec.rb]: You should also add a `context 'group has not reached maximum number of members'` with tests so that you have all permutations covered.

**Comment by dinever** [comment]
Thanks @johndbritton, just got the test case improved and will work on extracting logic into controllers later.

**Comment by johndbritton** [comment]
> will work on extracting logic into controllers later.

Let's leave that for a follow up pull request. I think @tarebyte is going to be introducing view models into the app.

**Comment by johndbritton** [comment]
This is good to :ship:

**Comment by dinever** [comment]
Hi Mark @tarebyte,

I wonder if there is anything more we need to do before getting this merged. :smile:

**Comment by tarebyte** [comment]
@dinever there shouldn't be, I'll pull this down tonight to give another once over before we ship it.

**Comment by dinever** [comment]
@tarebyte Sure. Thanks!

**Comment by johndbritton** [comment]
I just tried this out locally, it worked as advertised for me. Looks great!

**Comment by tarebyte** [comment]
@dinever sorry I merged a PR, if you merge master in one more time I'd really appreciate it

**Comment by dinever** [comment]
@tarebyte Just got this merged. :smile:

**Comment by johndbritton** [comment]
Thank you for the contribution @dinever :confetti_ball:

**Comment by drbeco** [comment]
Hello guys, quick question: in an ongoing assigment, is it ok to change the maximum number of members? Will it open spots in all teams and keep everything working flawless? Thank you!

**Comment by tarebyte** [comment]
> is it ok to change the maximum number of members? Will it open spots in all teams and keep everything working flawless? Thank you!

So everything will work as expected even if you change the number, we don't remove team members or anything.

**Comment by drbeco** [comment]
Thanks @tarebyte !
